### PR TITLE
feat(ccp)!: ccp splitting into async parts

### PR DESCRIPTION
**Pull request by monoid**
Make ccp more responsible with sepration of `CCProver` and async `CUGovernor`.  The `CCProver` keeps the state, and `CUGovernor` execute CU state changes under the `CCPRover`'s control.

### CODE REVIEW

File: ccp/src/prover.rs [hunk 1]
```diff
@@ -14,129 +14,81 @@
...
 const PROOF_DIR: &str = "cc_proofs";
 
 pub type CCResult<T> = Result<T, CCProverError>;
 
 pub struct CCProver {
-    cu_provers: HashMap<PhysicalCoreId, CUProver>,
-    cu_prover_config: CUProverConfig,
-    status: CCStatus,
+    governor_task: JoinHandle<()>,
+    governor_outlet: mpsc::Sender<GovernorMessage>,
```
Comment (raftedproc) [inline ccp/src/prover.rs:63]: Is it worth to alias `mpsc::Sender<GovernorMessage>` as a new type for semantic purposes?

Comment (monoid) [reply ccp/src/prover.rs:63]: Probably, yes.

File: ccp/src/prover/tests.rs [hunk 2]
```diff
@@ -97,17 +98,34 @@ fn load_state(state_dir: &Path) -> Option<CCPState> {
...
+async fn wait_for_proofs(
+    prover: &CCProver,
+    length: usize,
+    retries: usize,
+    retry_delay: Duration,
+) -> eyre::Result<Vec<CCProof>> {
+    for _retry in 0..retries {
+        tokio::time::sleep(retry_delay).await;
+        let proofs = prover.get_proofs_after("0".parse().unwrap()).await?;
+        if proofs.len() > length {
```
Comment (raftedproc) [inline ccp/src/prover/tests.rs:110]: Do you have a case in mind where `proofs.len() > length`?

File: ccp/src/prover/tests.rs [hunk 3]
```diff
@@ -97,17 +98,34 @@ fn load_state(state_dir: &Path) -> Option<CCPState> {
...
 }
 
+async fn wait_for_proofs(
+    prover: &CCProver,
+    length: usize,
+    retries: usize,
+    retry_delay: Duration,
+) -> eyre::Result<Vec<CCProof>> {
+    for _retry in 0..retries {
+        tokio::time::sleep(retry_delay).await;
```
[the remark this answers is not in the document]
Comment (monoid) [reply ccp/src/prover/tests.rs:108]: Yes, it is used _in tests only_ waiting for dataset generation.

File: ccp/src/prover/tests.rs [hunk 4]
```diff
@@ -97,17 +98,34 @@ fn load_state(state_dir: &Path) -> Option<CCPState> {
...
+        let proofs = prover.get_proofs_after("0".parse().unwrap()).await?;
```
Comment (raftedproc) [inline ccp/src/prover/tests.rs:109]: `get_proofs_after` first arg value is hardcoded here for a future potential parametrization?

File: ccp/src/prover/tests.rs (hunk 2 above)
Comment (monoid) [reply ccp/src/prover/tests.rs:110]: I don't understand what do you mean.

**Review by raftedproc — APPROVED** [review]
utACK

File: ccp/src/prover.rs [hunk 5]
```diff
@@ -14,129 +14,83 @@
...
-        self.utility_core_ids_handle.set_cores(utility_core_ids);
-    }
-}
-
-impl ToCCStatus for CCProver {
-    fn status(&self) -> CCStatus {
-        self.status
-    }
+    initial_msr_state: MSRState,
+    target_commitment: Option<Commitment>,
```
[the remark this answers is not in the document]
Comment (monoid) [reply ccp/src/prover.rs:73]: It just follows `on_active_commitment` method name: it contains both its arguments.  My idea is to rename it to `EpochAndAllocation`.

File: ccp/src/prover.rs [hunk 6]
```diff
@@ -237,129 +191,94 @@ impl CCProver {
...
+        Ok(wait_for
+            .await
+            .map_err(|_| CCProverError::UtilityClosedError)??)
     }
 
     #[inline]
-    pub async fn save_no_state(&self) -> CCResult<()> {
-        self.call_utility_save_state(None).await
+    pub async fn save_state(&self) -> CCResult<()> {
+        self.call_utility_save_state().await
```
[the remark this answers is not in the document]
Comment (monoid) [reply ccp/src/prover.rs:264]: These functions are public interface and implementation respectively.  Implementation name reflects how it is implemented.

**Pull request merged**